File: app/src/main/java/github/yaa110/memento/fragment/CategoryFragment.java
package github.yaa110.memento.fragment;

import android.content.Intent;
import android.view.View;

import com.google.android.material.snackbar.Snackbar;

import java.util.Locale;

import github.yaa110.memento.R;
import github.yaa110.memento.activity.NoteActivity;
import github.yaa110.memento.adapter.NoteAdapter;
import github.yaa110.memento.adapter.template.ModelAdapter;
import github.yaa110.memento.db.Controller;
import github.yaa110.memento.db.OpenHelper;
import github.yaa110.memento.fragment.template.RecyclerFragment;
import github.yaa110.memento.inner.Animator;
import github.yaa110.memento.model.DatabaseModel;
import github.yaa110.memento.model.Note;

public class CategoryFragment extends RecyclerFragment<Note, NoteAdapter> {
    public View protector;
    public View fab_type;
    public View fab_drawing;
    public boolean isFabOpen = false;

    private final ModelAdapter.ClickListener listener = new ModelAdapter.ClickListener() {
        @Override
        public void onClick(DatabaseModel item, int position) {
            startNoteActivity(item.type, item.id, position);
        }

        @Override
        public void onChangeSelection(boolean haveSelected) {
            toggleSelection(haveSelected);
        }

        @Override
        public void onCountSelection(int count) {
            onChangeCounter(count);
        }
    };

    public CategoryFragment() {
    }

    @Override
    public void init(View view) {
        protector = view.findViewById(R.id.protector);
        fab_type = view.findViewById(R.id.fab_type);
        fab_drawing = view.findViewById(R.id.fab_drawing);

        protector.setOnClickListener(view1 -> toggleFab(true));

        fab_type.setOnClickListener(view2 -> startNoteActivity(DatabaseModel.TYPE_NOTE_SIMPLE, DatabaseModel.NEW_MODEL_ID, 0));

        fab_drawing.setOnClickListener(view3 -> startNoteActivity(DatabaseModel.TYPE_NOTE_DRAWING, DatabaseModel.NEW_MODEL_ID, 0));
    }

    private void startNoteActivity(final int type, final long noteId, final int position) {
        toggleFab(true);

        new Thread() {
            @Override
            public void run() {
                try {
                    sleep(150);
                } catch (InterruptedException ignored) {
                }

                getActivity().runOnUiThread(() -> {
                    Intent intent = new Intent(getContext(), NoteActivity.class);
                    intent.putExtra(OpenHelper.COLUMN_TYPE, type);
                    intent.putExtra("position", position);
                    intent.putExtra(OpenHelper.COLUMN_ID, noteId);
                    intent.putExtra(OpenHelper.COLUMN_PARENT_ID, categoryId);
                    intent.putExtra(OpenHelper.COLUMN_THEME, categoryTheme);
                    startActivityForResult(intent, NoteActivity.REQUEST_CODE);
                });

                interrupt();
            }
        }.start();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, final Intent data) {
        if (requestCode == NoteActivity.REQUEST_CODE) {
            final int position = data.getIntExtra("position", 0);

            switch (resultCode) {
                case NoteActivity.RESULT_NEW:
                    Note note = new Note();
                    note.title = data.getStringExtra(OpenHelper.COLUMN_TITLE);
                    note.type = data.getIntExtra(OpenHelper.COLUMN_TYPE, DatabaseModel.TYPE_NOTE_SIMPLE);
                    note.createdAt = data.getLongExtra(OpenHelper.COLUMN_DATE, System.currentTimeMillis());
                    note.id = data.getLongExtra(OpenHelper.COLUMN_ID, DatabaseModel.NEW_MODEL_ID);
                    addItem(note, position);
                    break;
                case NoteActivity.RESULT_EDIT:
                    Note item = items.get(position);
                    item.title = data.getStringExtra(OpenHelper.COLUMN_TITLE);
                    refreshItem(position);
                    break;
                case NoteActivity.RESULT_DELETE:
                    new Thread() {
                        @Override
                        public void run() {
                            Controller.instance.deleteNotes(
                                    new String[]{
                                            String.format(Locale.US, "%d", data.getLongExtra(OpenHelper.COLUMN_ID, DatabaseModel.NEW_MODEL_ID))
                                    },
                                    categoryId
                            );

                            getActivity().runOnUiThread(() -> {
                                final Note deletedItem = deleteItem(position);
                                Snackbar.make(fab != null ? fab : selectionToolbar, "1 note was deleted", 7000)
                                        .setAction(R.string.undo, view -> new Thread() {
                                            @Override
                                            public void run() {
                                                Controller.instance.undoDeletion();
                                                Controller.instance.addCategoryCounter(deletedItem.categoryId, 1);

                                                getActivity().runOnUiThread(() -> addItem(deletedItem, position));

                                                interrupt();
                                            }
                                        }.start())
                                        .show();
                            });

                            interrupt();
                        }
                    }.start();
                    break;
            }
        }
    }

    @Override
    public void onClickFab() {
        toggleFab(false);
    }

    public void toggleFab(boolean forceClose) {
        if (isFabOpen) {
            isFabOpen = false;

            Animator.create(getContext())
                    .on(protector)
                    .setEndVisibility(View.GONE)
                    .animate(R.anim.fade_out);

            Animator.create(getContext())
                    .on(fab)
                    .animate(R.anim.fab_rotate_back);

            Animator.create(getContext())
                    .on(fab_type)
                    .setEndVisibility(View.GONE)
                    .animate(R.anim.fab_out);

            Animator.create(getContext())
                    .on(fab_drawing)
                    .setDelay(50)
                    .setEndVisibility(View.GONE)
                    .animate(R.anim.fab_out);
        } else if (!forceClose) {
            isFabOpen = true;

            Animator.create(getContext())
                    .on(protector)
                    .setStartVisibility(View.VISIBLE)
                    .animate(R.anim.fade_in);

            Animator.create(getContext())
                    .on(fab)
                    .animate(R.anim.fab_rotate);

            Animator.create(getContext())
                    .on(fab_type)
                    .setDelay(80)
                    .setStartVisibility(View.VISIBLE)
                    .animate(R.anim.fab_in);

            Animator.create(getContext())
                    .on(fab_drawing)
                    .setStartVisibility(View.VISIBLE)
                    .animate(R.anim.fab_in);
        }
    }

    @Override
    public int getLayout() {
        return R.layout.fragment_category;
    }

    @Override
    public String getItemName() {
        return "note";
    }

    @Override
    public Class<NoteAdapter> getAdapterClass() {
        return NoteAdapter.class;
    }

    @Override
    public ModelAdapter.ClickListener getListener() {
        return listener;
    }
}

File: app/src/main/java/github/yaa110/memento/fragment/CategoryFragment.kt
package github.yaa110.memento.fragment

import android.content.Intent
import android.view.View
import com.google.android.material.snackbar.Snackbar
import github.yaa110.memento.R
import github.yaa110.memento.activity.NoteActivity
import github.yaa110.memento.adapter.NoteAdapter
import github.yaa110.memento.adapter.template.ModelAdapter
import github.yaa110.memento.db.Controller
import github.yaa110.memento.db.OpenHelper
import github.yaa110.memento.fragment.template.RecyclerFragment
import github.yaa110.memento.inner.Animator
import github.yaa110.memento.model.DatabaseModel
import github.yaa110.memento.model.Note
import java.util.Locale

class CategoryFragment : RecyclerFragment<Note, NoteAdapter>() {
    lateinit var protector: View
    lateinit var fab_type: View
    lateinit var fab_drawing: View
    var isFabOpen: Boolean = false

    override val listener: ModelAdapter.ClickListener<*> =
        object : ModelAdapter.ClickListener<DatabaseModel> {
            override fun onClick(item: DatabaseModel, position: Int) {
                startNoteActivity(item!!.type, item.id, position)
            }

            override fun onChangeSelection(haveSelected: Boolean) {
                toggleSelection(haveSelected)
            }

            override fun onCountSelection(count: Int) {
                onChangeCounter(count)
            }
        }

    override fun init(view: View) {
        protector = view.findViewById(R.id.protector)
        fab_type = view.findViewById(R.id.fab_type)
        fab_drawing = view.findViewById(R.id.fab_drawing)

        protector.setOnClickListener { view1: View? -> toggleFab(true) }

        fab_type.setOnClickListener { view2: View? ->
            startNoteActivity(
                DatabaseModel.TYPE_NOTE_SIMPLE,
                DatabaseModel.NEW_MODEL_ID,
                0
            )
        }

        fab_drawing.setOnClickListener { view3: View? ->
            startNoteActivity(
                DatabaseModel.TYPE_NOTE_DRAWING,
                DatabaseModel.NEW_MODEL_ID,
                0
            )
        }
    }

    private fun startNoteActivity(type: Int, noteId: Long, position: Int) {
        toggleFab(true)

        object : Thread() {
            override fun run() {
                try {
                    sleep(150)
                } catch (ignored: InterruptedException) {
                }

                requireActivity().runOnUiThread {
                    val intent = Intent(context, NoteActivity::class.java)
                    intent.putExtra(OpenHelper.COLUMN_TYPE, type)
                    intent.putExtra("position", position)
                    intent.putExtra(OpenHelper.COLUMN_ID, noteId)
                    intent.putExtra(OpenHelper.COLUMN_PARENT_ID, categoryId)
                    intent.putExtra(OpenHelper.COLUMN_THEME, categoryTheme)
                    startActivityForResult(intent, NoteActivity.REQUEST_CODE)
                }

                interrupt()
            }
        }.start()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (data != null && requestCode == NoteActivity.REQUEST_CODE) {
            val position = data.getIntExtra("position", 0)

            when (resultCode) {
                NoteActivity.RESULT_NEW -> {
                    val note = Note()
                    note.title = data.getStringExtra(OpenHelper.COLUMN_TITLE)
                    note.type = data.getIntExtra(
                        OpenHelper.COLUMN_TYPE,
                        DatabaseModel.TYPE_NOTE_SIMPLE
                    )
                    note.createdAt = data.getLongExtra(
                        OpenHelper.COLUMN_DATE,
                        System.currentTimeMillis()
                    )
                    note.id = data.getLongExtra(
                        OpenHelper.COLUMN_ID,
                        DatabaseModel.NEW_MODEL_ID
                    )
                    addItem(note, position)
                }

                NoteActivity.RESULT_EDIT -> {
                    val item = items!![position]!!
                    item.title = data.getStringExtra(OpenHelper.COLUMN_TITLE)
                    refreshItem(position)
                }

                NoteActivity.RESULT_DELETE -> object : Thread() {
                    override fun run() {
                        Controller.instance!!.deleteNotes(
                            arrayOf(
                                String.format(
                                    Locale.US,
                                    "%d",
                                    data.getLongExtra(
                                        OpenHelper.COLUMN_ID,
                                        DatabaseModel.NEW_MODEL_ID
                                    )
                                )
                            ),
                            categoryId
                        )

                        requireActivity().runOnUiThread {
                            val deletedItem = deleteItem(position)
                            Snackbar.make(
                                (if (fab != null) fab else selectionToolbar),
                                "1 note was deleted",
                                7000
                            )
                                .setAction(R.string.undo) { view: View? ->
                                    object : Thread() {
                                        override fun run() {
                                            Controller.instance!!.undoDeletion()
                                            Controller.instance!!.addCategoryCounter(
                                                deletedItem!!.categoryId, 1
                                            )

                                            requireActivity().runOnUiThread {
                                                addItem(
                                                    deletedItem,
                                                    position
                                                )
                                            }

                                            interrupt()
                                        }
                                    }.start()
                                }
                                .show()
                        }

                        interrupt()
                    }
                }.start()
            }
        }
    }

    override fun onClickFab() {
        toggleFab(false)
    }

    fun toggleFab(forceClose: Boolean) {
        if (isFabOpen) {
            isFabOpen = false

            Animator.create(context)
                .on<View?>(protector)
                .setEndVisibility(View.GONE)
                .animate(R.anim.fade_out)

            Animator.create(context)
                .on(fab)
                .animate(R.anim.fab_rotate_back)

            Animator.create(context)
                .on<View?>(fab_type)
                .setEndVisibility(View.GONE)
                .animate(R.anim.fab_out)

            Animator.create(context)
                .on<View?>(fab_drawing)
                .setDelay(50)
                .setEndVisibility(View.GONE)
                .animate(R.anim.fab_out)
        } else if (!forceClose) {
            isFabOpen = true

            Animator.create(context)
                .on(protector)
                .setStartVisibility(View.VISIBLE)
                .animate(R.anim.fade_in)

            Animator.create(context)
                .on(fab)
                .animate(R.anim.fab_rotate)

            Animator.create(context)
                .on(fab_type)
                .setDelay(80)
                .setStartVisibility(View.VISIBLE)
                .animate(R.anim.fab_in)

            Animator.create(context)
                .on(fab_drawing)
                .setStartVisibility(View.VISIBLE)
                .animate(R.anim.fab_in)
        }
    }

    override val layout: Int
        get() = R.layout.fragment_category

    override val itemName: String
        get() = "note"

    override val adapterClass: Class<NoteAdapter>
        get() = NoteAdapter::class.java
}
